chanfitness/chanscore.go

```diff
@@ -0,0 +1,272 @@
+package chanfitness
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnpeer"
+)
+
+type eventType int
+
+const (
+	channelOpenedEvent eventType = iota
+	peerOnlineEvent
+	peerOfflineEvent
+	channelClosedEvent
+)
+
+func (e eventType) String() string {
+	switch e {
+	case channelOpenedEvent:
+		return "channel created"
+
+	case peerOnlineEvent:
+		return "peer online"
+
+	case peerOfflineEvent:
+		return "peer offline"
+
+	case channelClosedEvent:
+		return "channel closed"
+	}
+	return "unknown"
+}
+
+// channelEvent is a a timestamped event which is observed on a per channel basis.
+type channelEvent struct {
+	timestamp time.Time
+	eventType eventType
+}
+
+// ChannelScore represents the individual score for a single channel. Data
+// concerning the channel's lifetime (creation, peer uptime and close) are
+// stored in the channel event log, which is used to calculate scores such
+// as uptime and channel lifetime.
+type ChannelScore struct {
+	// id is the uInt64 of the short channel ID.
+	id uint64
+
+	peer [33]byte
+
+	// events is a log of timestamped events observed for the channel.
+	events []*channelEvent
+
+	firstSeen time.Time
+
+	// peerOnline is used to indicate whether we should be listening for peer
+	// offline or peer online notifications next.
+	peerOnline bool
+
+	// the quit channel should be used to terminate all goroutines running
+	// for the score.
+	quit chan struct{}
+
+	sync.Mutex
+
+	// monitoring is used as a temporary marker to indicate whether we are
+	// actively recording scores for a channel. It can be replaced when score
+	// data is persisted, because scores can be written to DB and removed
+	// entirely from memory (as is removing scores would delete all historical
+	// data, which would be wasteful).
+	monitoring bool
+}
+
+// MonitorPeer watches for events when the channel's remote peer goes offline
+// or online and records them in the channel score's event log. This function
+// is terminated when the channel is closed because we are only concerned with
+// peer behaviour across the lifetime of the channel.
+func (s *ChannelScore) MonitorPeer(online notifyOnline, offline notifyOffline) {
```

Comment:
Monitoring of peer uptime can be moved to the same level as channel close/opens. Each channel can have just a flat list of events: `[open, online, offline, online, offline, closed]` etc

Reply:
Actually, a bit of the complexity now seems to come from the fact that we are only monitoring peers we have channels with. I see no real reason we couldn't just monitor all our peers, since it is a superset of our channel peers. It would simplify the monitoring, since we could track the two independently.

To track all peers (regardless of whether they are a channel peer or not), we could add a new API similar to ChannelEvents, that would notify about peers going online/offline. This API would be useful in other places too.

Reply:
Not sure if I'm following what you mean by "flat list" of events? 

I'm going to go with storing offline/online on a per peer level as you suggested, so channels don't actually need to store any peer level events, and the EventStore can just get the range of peer events relevant to a channel based on its open/close time. 

In terms of storing an ordered set of peer events, is a []eventStruct{timestamp / type} like I currently have it ok?

Reply:
Actually, feels a little weird to have a set of channel events on the `chanEvent` level and then peer level events on the `EventStore` level which need to be combined when we need uptime/flat counts. 

I think I'm going to stick with everything on the channel event level, since there won't be much duplication (ie not expecting so many channels with the same peer that we'd be wasting tons of space).

Reply:
> In terms of storing an ordered set of peer events, is a []eventStruct{timestamp / type} like I currently have it ok?

Sounds good.